### DISCUSSION TITLE
PyInstaller not packaging Python Library (dlopen -> image not found).

### DISCUSSION THREAD

**Discussion by SimplyAhmazing** [pyinstaller/pyinstaller#5937] [Help]
I've packaged a [sample flask app](https://github.com/SimplyAhmazing/pyinstaller-tut) using PyInstaller but my OSX executable fails to run and shows the following executable,

```
Error loading Python lib '/Users/ahmed/Code/play/py-install-tut/dist/myscript.app/Contents/MacOS/Python': dlopen(/Users/ahmed/Code/play/py-install-tut/dist/myscript.app/Contents/MacOS/Python, 10): image not found
```

My guess is that PyInstaller is not packaging Python with my app. Here's what I ran,

``` bash
$ pyinstaller hello_flask.spec --onedir
83 INFO: PyInstaller: 3.2
83 INFO: Python: 3.4.3
87 INFO: Platform: Darwin-13.4.0-x86_64-i386-64bit
89 INFO: UPX is not available.
90 INFO: Extending PYTHONPATH with paths
['/Users/ahmed/Code/play/py-install-tut',
 '/Users/ahmed/Code/play/py-install-tut']
90 INFO: checking Analysis
99 INFO: checking PYZ
104 INFO: checking PKG
105 INFO: Building because toc changed
105 INFO: Building PKG (CArchive) out00-PKG.pkg
144 INFO: Bootloader /opt/boxen/pyenv/versions/3.4.3/Python.framework/Versions/3.4/lib/python3.4/site-packages/PyInstaller/bootloader/Darwin-64bit/run_d
144 INFO: checking EXE
145 INFO: Building because toc changed
145 INFO: Building EXE from out00-EXE.toc
145 INFO: Appending archive to EXE /Users/ahmed/Code/play/py-install-tut/build/hello_flask/hello_flask
155 INFO: Fixing EXE for code signing /Users/ahmed/Code/play/py-install-tut/build/hello_flask/hello_flask
164 INFO: checking COLLECT
WARNING: The output directory "/Users/ahmed/Code/play/py-install-tut/dist/hello_flask" and ALL ITS CONTENTS will be REMOVED! Continue? (y/n)y
1591 INFO: Removing dir /Users/ahmed/Code/play/py-install-tut/dist/hello_flask
1597 INFO: Building COLLECT out00-COLLECT.toc
2203 INFO: checking BUNDLE
WARNING: The output directory "/Users/ahmed/Code/play/py-install-tut/dist/myscript.app" and ALL ITS CONTENTS will be REMOVED! Continue? (y/n)y
3947 INFO: Removing dir /Users/ahmed/Code/play/py-install-tut/dist/myscript.app
3948 INFO: Building BUNDLE out00-BUNDLE.toc
3972 INFO: moving BUNDLE data files to Resource directory
```

When I open the contents of the packaged app in OSX I get the following files,

```
myscript.app/Contents/MacOS/
  _struct.cpython-34m.so
  hello_flask
  zlib.cpython-34m.so
```

When I double click the above `hello_flask` executable I get the following output in my terminal,

```
/Users/ahmed/Code/play/py-install-tut/dist/myscript.app/Contents/MacOS/hello_flask ; exit;                        
PyInstaller Bootloader 3.x
LOADER: executable is /Users/ahmed/Code/play/py-install-tut/dist/myscript.app/Contents/MacOS/hello_flask
LOADER: homepath is /Users/ahmed/Code/play/py-install-tut/dist/myscript.app/Contents/MacOS
LOADER: _MEIPASS2 is NULL
LOADER: archivename is /Users/ahmed/Code/play/py-install-tut/dist/myscript.app/Contents/MacOS/hello_flask
LOADER: Extracting binaries
LOADER: Executing self as child
LOADER: set _MEIPASS2 to /Users/ahmed/Code/play/py-install-tut/dist/myscript.app/Contents/MacOS
PyInstaller Bootloader 3.x
LOADER: executable is /Users/ahmed/Code/play/py-install-tut/dist/myscript.app/Contents/MacOS/hello_flask
LOADER: homepath is /Users/ahmed/Code/play/py-install-tut/dist/myscript.app/Contents/MacOS
LOADER: _MEIPASS2 is /Users/ahmed/Code/play/py-install-tut/dist/myscript.app/Contents/MacOS
LOADER: archivename is /Users/ahmed/Code/play/py-install-tut/dist/myscript.app/Contents/MacOS/hello_flask
LOADER: Already in the child - running user's code.
LOADER: Python library: /Users/ahmed/Code/play/py-install-tut/dist/myscript.app/Contents/MacOS/Python
Error loading Python lib '/Users/ahmed/Code/play/py-install-tut/dist/myscript.app/Contents/MacOS/Python': dlopen(/Users/ahmed/Code/play/py-install-tut/dist/myscript.app/Contents/MacOS/Python, 10): image not found
LOADER: Back to parent (RC: 255)
LOADER: Doing cleanup
LOADER: Freeing archive status for /Users/ahmed/Code/play/py-install-tut/dist/myscript.app/Contents/MacOS/hello_flask

[Process completed]
```

I've also tried running this on a co-workers mac OSX and I get the same issue.

**Comment by markgukov** [comment]
Experiencing a similar issue.

**Comment by bdeluca** [comment]
+1

**Comment by SHxKM** [comment]
Same here.

**Comment by jessicaabreu** [comment]
+1

**Comment by htgoebel** [comment]
Did you check whether you python executable is linked statically?

**Comment by seannaman** [comment]
I realize this is an old thread but I'm having the same issue. The build appears fine w/ no errors but this error occurs when trying to open the executable. Did anyone find a recipe to resolve this?

**Comment by personalnadir** [comment]
Likewise, I've installed Python 3.7 using Mac Ports. No errors have jumped out at me from the output. Has anyone managed to find workarounds, fixes or reasons why this might be happening?

**Comment by personalnadir** [comment]
I cleaned the config and cache and now it runs fine. However on reflection I think I might have been trying to run from the build directory as opposed to the dist directory. Sadly I can't confirm which was the case as the attempts aren't in my shell history.

**Comment by seannaman** [comment]
My issue was ultimately related to Pyinstaller being unable to locate a .ui file that my program used. Now everything works. This was not at all obvious from the error message.

**Comment by DanielAron** [comment]
+1

**Comment by ddnomad** [comment]
Same issue :|

**Comment by coder-lkl** [comment]
The same :(
Did anyone find reasons?

**Comment by theguymeyer** [comment]
+1

getting this when running from the build folder...

`[59556] Error loading Python lib '/Users/guy/development/python3/pyqt/build/mainwindow/Python': dlopen: dlopen(/Users/guy/development/python3/pyqt/build/mainwindow/Python, 10): image not found`

**Comment by rpalri** [comment]
I have the same issue. Anyone with a solution?

**Comment by noisyboiler** [comment]
another one here.

**Comment by BoboTiG** [comment]
The problem here is you need to pass `coll` to `BUNDLE()`, not `exe`.

So just replace `exe` with `coll` [here](https://github.com/SimplyAhmazing/pyinstaller-tut/blob/6d4ada83b5744c4eb4e14cd5db2535c08af1a274/foo.spec#L36) and [here](https://github.com/SimplyAhmazing/pyinstaller-tut/blob/6d4ada83b5744c4eb4e14cd5db2535c08af1a274/hello_flask.spec#L35), and it will work :)

**Reply by pklapperich** [reply]
Is this a documentation bug, then? The documentation example shows using exe:
```
app = BUNDLE(exe,
         name='myscript.app',
         icon=None,
         bundle_identifier=None)
```

https://pyinstaller.readthedocs.io/en/stable/spec-files.html#spec-file-options-for-a-mac-os-x-bundle

**Reply by rokm** [reply]
That's an example for a `onefile` app bundle. In that case, there's no `coll` and `exe` needs to be passed to `BUNDLE()`. 

The referenced [spec](https://github.com/SimplyAhmazing/pyinstaller-tut/blob/6d4ada83b5744c4eb4e14cd5db2535c08af1a274/hello_flask.spec#L35) however, was for `onedir` app bundle, in which case `coll` needs to be passed to `BUNDLE()`. But yeah, the docs should be updated to explain that...

**Comment by fishbacp** [comment]
@BoboTiG : Thank you so much! Spent two days trying to fix this!

**Comment by SpecialCharacter** [comment]
It is still an issue in 2023.
macOS Catalina 10.15.7
